File: data_structure/stack/10828_GGM.py
# -*- coding: utf-8 -*-
# 정수를 저장하는 스택을 구현한 다음, 입력으로 주어지는 명령을 처리하는 프로그램을 작성하시오.

from array import array
from sys import stdin, stdout


class Stack:
    def __init__(self):

        self._arr = array("i", [0] * 10001)  # i == sint16
        self._top = -1

    def push(self, x):
        self._top += 1
        self._arr[self._top] = x

    def pop(self):
        if self.is_empty():  # 편의를 위해 empty 미사용
            stdout.write("-1" + "\n")
            return

        self.top()
        self._top -= 1

    def size(self):
        stdout.write(str(self._top + 1) + "\n")

    def empty(self):
        stdout.write(str(int(self._top == -1)) + "\n")

    def is_empty(self):
        return self._top == -1

    def top(self):
        if self.is_empty():  # 편의를 위해 empty 미사용
            stdout.write("-1" + "\n")
            return

        stdout.write(str(self._arr[self._top]) + "\n")


if __name__ == "__main__":
    MIS = lambda x: x.split(" ")
    N = int(stdin.readline())
    s = Stack()

    for _ in range(N):
        command = MIS(stdin.readline().rstrip("\n"))

        if command[0] == "push":
            sint = int(command[1])
            getattr(s, command[0])(sint)
            continue

        getattr(s, command[0])()

File: data_structure/stack/10828_GGM.js
class Stack {
    constructor() {
        // Javascript arrays are typically implemented as hashmaps with one added feature:
        // there is an attribute `length`
        this._arr = new Array(10001);
        this._top = -1;
    }

    push(x) {
        this._top += 1;
        this._arr[this._top] = x;
    }
    pop() {
        if (this.is_empty()) {
            console.log("-1");
            return;
        }
        this.top();
        this._top -= 1;
    }
    size() {
        console.log(String(this._top + 1));
    }
    empty() {
        var temp;
        if (this._top == -1) {
            temp = "1";
        } else {
            temp = "0";
        }
        console.log(temp);
    }
    is_empty() {
        return this._top == -1;
    }
    top() {
        if (this.is_empty()) {
            console.log(" -1");
            return;
        }
        console.log(this._arr[this._top]);
    }
}

function main() {
    // 입력 받아오는 부분은 정답 코드 참고함
    let input = require("fs").readFileSync("./input").toString().split("\n");
    // let input = require("fs").readFileSync("/dev/stdin").toString().split("\n");
    let n = +input.shift();
    let s = new Stack();
    for (let i = 0; i < n; ++i) {
        command = input[i];
        switch (command) {
            case "pop":
                s.pop();
                break;
            case "size":
                s.size();
                break;
            case "empty":
                s.empty();
                break;
            case "top":
                s.top();
                break;
            default:
                s.push(+command.substring(5));
        }
    }
}

main();
